### PR TITLE
Change NumericToken from IDataToken<decimal> to IDataToken<double> and fix #765

### LINKED ISSUE #765: NullReferenceException in GetPages() ... ResourceStore.GetFont()
Using this file: https://arxiv.org/pdf/2108.11480.pdf

Nuget package: [0.1.9-alpha-20240128-f886e](https://www.nuget.org/packages/PdfPig/0.1.9-alpha-20240128-f886e)

Simply opening this PDF, and asking for the pages is throwing an exception getting a font, it appears.

```
using var pdoc = PdfDocument.Open(content.Stream, new ParsingOptions { SkipMissingFonts = true, UseLenientParsing = true });
var docpages = pdoc.GetPages().ToList();
```

Getting this exception thrown:
```
   at System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at UglyToad.PdfPig.Util.StackDictionary`2.TryGetValue(K key, V& result)
   at UglyToad.PdfPig.Content.ResourceStore.GetFont(NameToken name)
   at UglyToad.PdfPig.Graphics.BaseStreamProcessor`1.ShowText(IInputBytes bytes)
   at UglyToad.PdfPig.Graphics.Operations.TextShowing.ShowText.Run(IOperationContext operationContext)
   at UglyToad.PdfPig.Graphics.BaseStreamProcessor`1.ProcessOperations(IReadOnlyList`1 operations)
   at UglyToad.PdfPig.Graphics.BaseStreamProcessor`1.ProcessFormXObject(StreamToken formStream, NameToken xObjectName)
   at UglyToad.PdfPig.Graphics.BaseStreamProcessor`1.ApplyXObject(NameToken xObjectName)
   at UglyToad.PdfPig.Graphics.Operations.InvokeNamedXObject.Run(IOperationContext operationContext)
   at UglyToad.PdfPig.Graphics.BaseStreamProcessor`1.ProcessOperations(IReadOnlyList`1 operations)
   at UglyToad.PdfPig.Graphics.ContentStreamProcessor.Process(Int32 pageNumberCurrent, IReadOnlyList`1 operations)
   at UglyToad.PdfPig.Parser.PageFactory.ProcessPage(Int32 pageNumber, DictionaryToken dictionary, NamedDestinations namedDestinations, MediaBox mediaBox, CropBox cropBox, UserSpaceUnit userSpaceUnit, PageRotationDegrees rotation, TransformationMatrix initialMatrix, IReadOnlyList`1 operations)
   at UglyToad.PdfPig.Content.BasePageFactory`1.Create(Int32 number, DictionaryToken dictionary, PageTreeMembers pageTreeMembers, NamedDestinations namedDestinations)
   at UglyToad.PdfPig.Content.Pages.GetPage[TPage](IPageFactory`1 pageFactory, Int32 pageNumber, NamedDestinations namedDestinations, ParsingOptions parsingOptions)
   at UglyToad.PdfPig.Content.Pages.GetPage(Int32 pageNumber, NamedDestinations namedDestinations, ParsingOptions parsingOptions)
   at UglyToad.PdfPig.PdfDocument.GetPage(Int32 pageNumber)
   at UglyToad.PdfPig.PdfDocument.<GetPages>d__34.MoveNext()
```

### PR DESCRIPTION
**Pull request by BobLd**
### Change NumericToken from IDataToken<decimal> to IDataToken<double> and fix #765

First, I'm sorry for this very long PR as I'm sure it won't be easy to review.

The main purpose of the PR is to switch from `IDataToken<decimal>` to `IDataToken<double>` for `NumericToken` base interface. This is in order to fix issue #765. The actual fix is located in the `NumericTokenizer` on line 156, where we now can use `double.TryParse(...)` instead of `decimal.TryParse(...)` which was failing when trying to parse the following value into a `decimal`:

> 1797693134862315631259457270791380001216713246039491126358296914738626428928

Switching to double solves the bug and the pages can now be accessed without bug.

I've made the changes manually and I've limited them to fixing the impact of switching to `IDataToken<double>`. The code base still contains many `decimal` values that will need to be addressed at a later point.

Do let me know if you think, but it reinforces my opinion that `decimal` do not belong in the pdf world.

### CODE REVIEW

File: src/UglyToad.PdfPig.Tokens/NumericToken.cs [hunk 1]
```diff
@@ -155,13 +156,32 @@ public class NumericToken : IDataToken<decimal>
...
         /// The value of this number as a <see langword="double"/>.
         /// </summary>
-        public double Double => (double)Data;
+        public double Double => Data;
 
         /// <summary>
         /// Create a <see cref="NumericToken"/>.
         /// </summary>
         /// <param name="value">The number to represent.</param>
+        //[Obsolete("Use double constructor instead.")]
```
Comment (BobLd) [inline src/UglyToad.PdfPig.Tokens/NumericToken.cs:165]: I can't run the tests if I mark this method obsolete... Maybe there are some rules somewhere that prevent that...

File: src/UglyToad.PdfPig.Tokenization/NumericTokenizer.cs [hunk 2]
```diff
@@ -153,7 +153,7 @@ public bool TryTokenize(byte currentByte, IInputBytes inputBytes, out IToken tok
                         token = NumericToken.OneThousand;
                         return true;
                     default:
-                        if (!decimal.TryParse(str, NumberStyles.Any, CultureInfo.InvariantCulture, out var value))
+                        if (!double.TryParse(str, NumberStyles.Any, CultureInfo.InvariantCulture, out var value))
```
Comment (EliotJones) [inline src/UglyToad.PdfPig.Tokenization/NumericTokenizer.cs:156]: Random thought while skimming this, is it possible `double` supports/returns `NaN` where previously decimal wouldn't? I think the integration tests would catch that tbh so it's probably fine.

File: src/UglyToad.PdfPig.Tokens/NumericToken.cs (hunk 1 above)
Comment (EliotJones) [reply src/UglyToad.PdfPig.Tokens/NumericToken.cs:165]: Is it still used after this changeset or could it be removed in a follow-up PR? I think the next version bump will be quite breaking anyway so in for a penny in for a pound, just remove it if unneeded internally

File: src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs [hunk 3]
```diff
@@ -944,7 +944,7 @@ public CalRGBColorSpaceDetails([NotNull] IReadOnlyList<double> whitePoint, [CanB
         }
 
         /// <summary>
-        /// Transforms the supplied ABC color to RGB (sRGB) using the propties of this <see cref="CalRGBColorSpaceDetails"/>
+        /// Transforms the supplied ABC color to RGB (sRGB) using the propeties of this <see cref="CalRGBColorSpaceDetails"/>
```
Comment (EliotJones) [inline src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs:947]:
```suggestion
        /// Transforms the supplied ABC color to RGB (sRGB) using the properties of this <see cref="CalRGBColorSpaceDetails"/>

```

**Review by EliotJones — COMMENTED** [review]
I'm halfway through, I'll try to get the rest reviewed soon

File: src/UglyToad.PdfPig.Tokenization/NumericTokenizer.cs (hunk 2 above)
Comment (BobLd) [reply src/UglyToad.PdfPig.Tokenization/NumericTokenizer.cs:156]: I think you're right. I'm pretty sure that is `str` = "NaN", then `double.TryParse()` would return `NaN`. But as you said, I think we are good for the moment as the integration tests would have caught any issue.

Good to keep in mind that in can happen though

File: src/UglyToad.PdfPig.Tokens/NumericToken.cs (hunk 1 above)
Comment (BobLd) [reply src/UglyToad.PdfPig.Tokens/NumericToken.cs:165]: agreed yes, I'll remove that in a follow up PR

**Comment by BobLd** [comment]
@EliotJones just an FYI, it seems like your review was reset when I pushed the typo fix

**Review by EliotJones — APPROVED** [review]

**Pull request merged**